trustfall_derive/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+                Ok(syn::parse_quote! {
+                    pub(crate) fn #conversion_name(&self) -> Option<#final_type> {
+                        match self {
+                            Self::#variant_ident { #fields } => Some((#fields)),
+                            _ => None,
+                        }
+                    }
+                })
+            }
+        }
+        syn::Fields::Unnamed(tuple_fields) if !tuple_fields.unnamed.is_empty() => {
+            if tuple_fields.unnamed.len() == 1 {
+                // Tuple variants with only a single field return `Option<&ThatField>`.
+                let field_type = &tuple_fields.unnamed[0].ty;
+                Ok(syn::parse_quote! {
+                    pub(crate) fn #conversion_name(&self) -> Option<&#field_type> {
+                        match self {
+                            Self::#variant_ident(x) => Some(x),
+                            _ => None,
+                        }
+                    }
+                })
+            } else {
+                // Tuple variants with multiple fields return
+                // `Option<(&FirstField, &SecondField, ...)>`.
+                let final_type: proc_macro2::TokenStream =
+                    tuple_of_field_types(&tuple_fields.unnamed);
+                let mut fields = syn::punctuated::Punctuated::<_, syn::Token![,]>::new();
+                for (i, _) in tuple_fields.unnamed.iter().enumerate() {
+                    fields.push(quote::format_ident!("x{i}"));
+                }
+                Ok(syn::parse_quote! {
+                    pub(crate) fn #conversion_name(&self) -> Option<#final_type> {
+                        match self {
+                            Self::#variant_ident(#fields) => Some((#fields)),
+                            _ => None,
+                        }
+                    }
+                })
+            }
+        }
+        _ => {
+            // Either unit variant, or fieldless struct/tuple variant.
+            Ok(syn::parse_quote! {
+                pub(crate) fn #conversion_name(&self) -> Option<()> {
+                    match self {
+                        Self::#variant_ident => Some(()),
+                        _ => None,
+                    }
+                }
+            })
+        }
+    }
+}
+
+/// Returns a tuple of references to all field types.
+/// The input must contain more than one field.
+fn tuple_of_field_types(
+    fields: &Punctuated<syn::Field, syn::Token![,]>,
+) -> proc_macro2::TokenStream {
+    if fields.len() > 1 {
+        let mut punct = syn::punctuated::Punctuated::<_, syn::Token![,]>::new();
+        for field in fields.iter() {
+            let ty = &field.ty;
+            punct.push(quote!(&#ty))
+        }
+        quote!((#punct))
+    } else {
+        panic!(
+            "list of fields had {} field(s), which is not more than one field",
+            fields.len()
+        );
+    }
+}
+
+fn to_lower_snake_case(value: &str) -> String {
```

Comment:
Perhaps some tests will help verify this. It looks right to me, but it would be nice with some tests for sketchy names, like starting with uppercase, double uppercase, etc.

I saw that there are integration tests for it, but I went looking for them in this file

Reply:
It's easy enough to add them here too, I don't mind 👍